File: LeetCode/022.Generate Parentheses/Solution.java
public class Solution {
  public List<String> generateParenthesis(int n) {
    List<String> result = new ArrayList<String>();

    helper(result, "", n, n);
    return result;
  }

  private void helper(List<String> result, String s, int left, int right) {
    if (left > right || left < 0 || right < 0) {
      return;
    }

    if (left == 0 && right == 0) {
      result.add(s);
    }

    helper(result, s + "(", left - 1, right);
    helper(result, s + ")", left, right - 1);
  }
}

File: LeetCode/022.Generate Parentheses/Solution.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 0:
            return list()
        ans = list()
        self.helper(n, n, "", ans)
        return ans

    def helper(self, left, right, taken, ans):
        if (left > right or left < 0 or right < 0):
            return
        if left == 0 and right == 0:
            ans.append(taken)

        self.helper(left - 1, right, taken + '(', ans)
        self.helper(left, right - 1, taken + ')', ans)
